Create embedded schema guide (#3761)
elixir-ecto/ecto: guides/introduction/Embedded Schemas.md

They're great for:

- Storing additional data about an entity without modifying your schema.
- Storing additional data about an entity without tedious foreign key/many-to-many relationships
- Validating arbitrary data structures

Example use cases:

- User profiles: storing information like profile pictures, settings
- Shop products: storing additional product images

While embedded schemas require an initial migration to create the field, any subsequent modifications to the data structure don't.

This feature is great for situations where you'd like to store data associated to a specific entity, but a relation or new table would be overkill. You can even write changesets to perform validations on this data!

## Example

A common use of embedded schemas is to store additional information about the current user. The finished example for this guide's embedded schema would look like this:

```elixir
defmodule User do
  use Ecto.Schema

  schema "users" do
    field :is_active, :boolean
    field :email, :string
    field :confirmed_at, :naive_datetime

> 

Reviewer: ```suggestion
```

    embeds_one "profile" do
      field :age, :integer
      field :gender, :integer
      field :avatar_url, :string
    end

    timestamps()
  end
end
```

Let's kick things off with a tutorial to explain how to recreate this use case: building a solution for user profiles in an example app.

### Writing the schema

> The first step is to write the structure of the data we're storing. In the case of our profile, we'd like to store the user's age, gender and a profile picture.

Reviewer: ```suggestion
The first step is to write the structure of the data we're storing. In the case of our profile, we'd like to store the user's age, favorite color, and a profile picture.
```

To begin writing this embedded schema, we must first think about what structure we want. Do we want to store an array of structures (using `embeds_many`) or just one (using `embeds_one`)?

In this case, every user should have only one profile associated to them, so we'll begin by writing like any other Ecto schema:

```elixir
defmodule User do
  use Ecto.Schema

  schema "users" do
    field :is_active, :boolean
    field :email, :string
    field :confirmed_at, :naive_datetime

    embeds_one :profile do
      field :age, :integer

>       field :gender, :integer

Reviewer: ```suggestion
      field :favorite_color, Ecto.Enum, values: [:red, :green, :blue, :pink, :black, :orange]
```

      field :avatar_url, :string
    end

    timestamps()
  end
end
```

We can, however, clean this up a little. You can separate the profile to a distinct module and keep the `User` module tidy:

```elixir
# user/user.ex
defmodule User do
  use Ecto.Schema

  schema "users" do
    field :is_active, :boolean
    field :email, :string
    field :confirmed_at, :naive_datetime

    embeds_one :profile, UserProfile

    timestamps()
  end
end

# user/user_profile.ex
defmodule UserProfile do
  embedded_schema "profile" do
    field :age, :integer

>     field :gender, :integer

Reviewer: ```suggestion
    field :favorite_color, Ecto.Enum, values: [:red, :green, :blue, :pink, :black, :orange]
```
